// erela.js/dist/lib/structures/Player.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Queue_1 = require("./Queue");
/**
 * The Player class.
 */
class Player {
    /**
     * Creates an instance of Player.
     * @param {ErelaClient} erela - The Erela client.
     * @param {Node} node - The Erela Node.
     * @param {IPlayerOptions} options - The player options.
     * @param {object} extra - Extra data to pass when extending for custom classes.
     */
    constructor(erela, node, options, extra) {
        this.erela = erela;
        this.node = node;
        this.options = options;
        this.guild = options.guild;
        this.textChannel = options.textChannel;
        this.voiceChannel = options.voiceChannel;
        this.bands = [...new Array(14)].map((_, i) => JSON.parse(`{"band": ${i}, "gain": 0.0}`));
        this.queue = new Queue_1.Queue(erela);
        this.playing = false;
        this.position = 0;
        this.volume = 100;
        this.trackRepeat = false;
        this.queueRepeat = false;
    }
    /**
     * Plays the next track in the queue.
     */
    play() {
        if (this.queue[0]) {
            this.playing = true;
            this.node.send({
                op: "play",
                guildId: this.guild.id,
                track: this.queue[0].track,
            });
            this.erela.emit("trackStart", this, this.queue[0]);
        }
    }
    /**
     * Sets the players volume.
     * @param {number} volume - The volume to set.
     */
    setVolume(volume) {
        if (isNaN(volume)) {
            throw new RangeError("Player#setVolume(volume: number) Volume must be a number");
        }
        if (volume <= 0 || volume > 100) {
            throw new RangeError("Player#setVolume(volume: number) Volume can not be lower than 1 or higher than 100, must be or be between 1 and 100.");
        }
        this.volume = volume;
        this.node.send({
            op: "volume",
            guildId: this.guild.id,
            volume,
        });
    }
    /**
     * Sets the players equalizer. Pass a empty array to reset the bands.
     * @param {Array<EqualizerBand>} bands - The array of bands to set.
     * @example
     * player.setEQ([
     *      { band: 0, gain: 0.15 },
     *      { band: 1, gain: 0.15 },
     *      { band: 2, gain: 0.15 }
     * ]);
     */
    setEQ(bands) {
        if (!bands || !Array.isArray(bands)) {
            throw new RangeError("Player#setEQ(bands: IEqualizerBand[]) Bands must be a array of bands.");
        }
        if (bands.length === 0) {
            this.bands.map((v, i) => this.bands[i].gain = 0.0);
        }
        else {
            bands.forEach(({ band, gain }) => this.bands[band].gain = gain);
        }
        this.node.send({
            op: "equalizer",
            guildId: this.guild.id,
            bands: this.bands.map((band) => band),
        });
    }
    /**
     * Sets the track repeat.
     * @param {boolean} repeat - If track repeat should be enabled.
     */
    setTrackRepeat(repeat) {
        if (typeof repeat !== "boolean") {
            throw new RangeError("Player#setTrackRepeat(repeat: boolean) Repeat can only be \"true\" or \"false\".");
        }
        if (repeat) {
            this.trackRepeat = true;
            this.queueRepeat = false;
        }
        else {
            this.trackRepeat = false;
            this.queueRepeat = false;
        }
    }
    /**
     * Sets the queue repeat.
     * @param {boolean} repeat - If queue repeat should be enabled.
     */
    setQueueRepeat(repeat) {
        if (typeof repeat !== "boolean") {
            throw new RangeError("Player#setQueueRepeat(repeat: boolean) Repeat can only be \"true\" or \"false\".");
        }
        if (repeat) {
            this.trackRepeat = false;
            this.queueRepeat = true;
        }
        else {
            this.trackRepeat = false;
            this.queueRepeat = false;
        }
    }
    updateState(data) {
        this.position = data.position || 0;
    }
    /**
     * Stops the current track.
     */
    stop() {
        this.node.send({
            op: "stop",
            guildId: this.guild.id,
        });
    }
    /**
     * Pauses the current track.
     * @param {boolean} pause - Whether to pause the current track.
     */
    pause(pause) {
        if (typeof pause !== "boolean") {
            throw new RangeError("Player#pause(pause: boolean) Pause can only be \"true\" or \"false\".");
        }
        this.playing = !pause;
        this.node.send({
            op: "pause",
            guildId: this.guild.id,
            pause,
        });
    }
    /**
     * Seeks to the position in the current track.
     * @param {boolean} pause - Whether to pause the current track.
     */
    seek(position) {
        if (!this.queue[0]) {
            throw new RangeError("Player#seek(position: number) Can only seek when theres a track in the queue.");
        }
        if (isNaN(position)) {
            throw new RangeError("Player#seek(position: number) Position must be a number.");
        }
        if (position < 0 || position > this.queue[0].duration) {
            throw new RangeError(`Player#seek(position: number) Position can not be smaller than 0 or bigger than ${this.queue[0].duration}.`);
        }
        this.position = position;
        this.node.send({
            op: "seek",
            guildId: this.guild.id,
            position,
        });
    }
}
exports.Player = Player;
