// erela.js/dist/lib/structures/Track.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * The Track class.
 * @export
 * @class Track
 */
class Track {
    /**
     * Creates an instance of Track.
     * @param {ITrackData} data - The data to pass.
     * @param {User} user - The user who requested the track.
     */
    constructor(data, user) {
        if (!data || !user) {
            throw new RangeError("Track constructor must have all two parameters filled.");
        }
        try {
            this.track = data.track;
            this.identifier = data.info.identifier;
            this.isSeekable = data.info.isSeekable;
            this.author = data.info.author;
            this.duration = data.info.length;
            this.isStream = data.info.isStream;
            this.title = data.info.title;
            this.uri = data.info.uri;
            this.thumbnail = this.uri.includes("youtube") ?
                `https://img.youtube.com/vi/${this.identifier}/default.jpg` : "";
            this.requester = user;
        }
        catch (err) {
            throw new RangeError(`Invalid track passed. Reason: ${err}`);
        }
    }
}
exports.Track = Track;
