// erela.js/src/lib/structures/Player.ts
import { Guild, TextChannel, VoiceChannel } from "discord.js";
import { ErelaClient } from "../ErelaClient";
import { Node } from "./Node";
import { Queue } from "./Queue";

/**
 * The IPlayerOptions interface.
 */
export interface IPlayerOptions {
    /**
     * The guild to connect to.
     */
    guild: Guild;
    /**
     * The text channel to connect to.
     */
    textChannel: TextChannel;
    /**
     * The voice channel to connect to.
     */
    voiceChannel: VoiceChannel;
    /**
     * Whether to deafen the client.
     */
    selfDeaf?: boolean;
    /**
     * Whether to mute the client.
     */
    selfMute?: boolean;
}

/**
 * The IEqualizerBand interface.
 */
export interface IEqualizerBand {
    /**
     * The band for the equalizer band.
     */
    band: number;
    /**
     * The gain for the equalizer band.
     */
    gain: number;
}

/**
 * The Player class.
 */
export class Player {
    public readonly erela: ErelaClient;
    /**
     * The players node.
     */
    public readonly node: Node;
    /**
     * The players options.
     */
    public readonly options: IPlayerOptions;
    /**
     * The players guild it's connected to.
     */
    public readonly guild: Guild;
    /**
     * The players text channel it's connected to.
     */
    public readonly textChannel: TextChannel;
    /**
     * The players voice channel it's connected to.
     */
    public voiceChannel: VoiceChannel;
    /**
     * The players equalizer bands.
     */
    public bands: IEqualizerBand[];
    /**
     * The players queue.
     */
    public readonly queue: Queue;
    /**
     * Whether the player is playing.
     */
    public playing: boolean;
    /**
     * The players current position in the track.
     */
    public position: number;
    /**
     * The players current volume.
     */
    public volume: number;
    /**
     * Whether the player is repeating the current track.
     */
    public trackRepeat: boolean;
    /**
     * Whether the player is repeating the queue.
     */
    public queueRepeat: boolean;

    /**
     * Creates an instance of Player.
     * @param {ErelaClient} erela - The Erela client.
     * @param {Node} node - The Erela Node.
     * @param {IPlayerOptions} options - The player options.
     * @param {object} extra - Extra data to pass when extending for custom classes.
     */
    public constructor(erela: ErelaClient, node: Node, options: IPlayerOptions, extra: object) {
        this.erela = erela;
        this.node = node;
        this.options = options;
        this.guild = options.guild;
        this.textChannel = options.textChannel;
        this.voiceChannel = options.voiceChannel;
        this.bands = [...new Array(14)].map((_, i) => JSON.parse(`{"band": ${i}, "gain": 0.0}`));
        this.queue = new Queue(erela);
        this.playing = false;
        this.position = 0;
        this.volume = 100;
        this.trackRepeat = false;
        this.queueRepeat = false;
    }

    /**
     * Plays the next track in the queue.
     */
    public play(): void {
        if (this.queue[0]) {
            this.playing = true;
            this.node.send({
                op: "play",
                guildId: this.guild.id,
                track: this.queue[0].track,
            });
            this.erela.emit("trackStart", this, this.queue[0]);
        }
    }

    /**
     * Sets the players volume.
     * @param {number} volume - The volume to set.
     */
    public setVolume(volume: number): void {
        if (isNaN(volume)) {
            throw new RangeError("Player#setVolume(volume: number) Volume must be a number");
        }
        if (volume <= 0 || volume > 100) {
            throw new RangeError("Player#setVolume(volume: number) Volume can not be lower than 1 or higher than 100, must be or be between 1 and 100.");
        }
        this.volume = volume;
        this.node.send({
            op: "volume",
            guildId: this.guild.id,
            volume,
        });
    }

    /**
     * Sets the players equalizer. Pass a empty array to reset the bands.
     * @param {Array<EqualizerBand>} bands - The array of bands to set.
     * @example
     * player.setEQ([
     *      { band: 0, gain: 0.15 },
     *      { band: 1, gain: 0.15 },
     *      { band: 2, gain: 0.15 }
     * ]);
     */
    public setEQ(bands: IEqualizerBand[]): void {
        if (!bands || !Array.isArray(bands)) {
            throw new RangeError("Player#setEQ(bands: IEqualizerBand[]) Bands must be a array of bands.");
        }
        if (bands.length === 0) {
            this.bands.map((v, i) => this.bands[i].gain = 0.0);
        } else {
            bands.forEach(({ band, gain }) => this.bands[band].gain = gain);
        }
        this.node.send({
            op: "equalizer",
            guildId: this.guild.id,
            bands: this.bands.map((band) => band),
        });
    }

    /**
     * Sets the track repeat.
     * @param {boolean} repeat - If track repeat should be enabled.
     */
    public setTrackRepeat(repeat: boolean): void {
        if (typeof repeat !== "boolean") { throw new RangeError("Player#setTrackRepeat(repeat: boolean) Repeat can only be \"true\" or \"false\"."); }
        if (repeat) {
            this.trackRepeat = true;
            this.queueRepeat = false;
        } else {
            this.trackRepeat = false;
            this.queueRepeat = false;
        }
    }

    /**
     * Sets the queue repeat.
     * @param {boolean} repeat - If queue repeat should be enabled.
     */
    public setQueueRepeat(repeat: boolean): void {
        if (typeof repeat !== "boolean") { throw new RangeError("Player#setQueueRepeat(repeat: boolean) Repeat can only be \"true\" or \"false\"."); }
        if (repeat) {
            this.trackRepeat = false;
            this.queueRepeat = true;
        } else {
            this.trackRepeat = false;
            this.queueRepeat = false;
        }
    }

    public updateState(data: any): void {
        this.position = data.position || 0;
    }

    /**
     * Stops the current track.
     */
    public stop(): void {
        this.node.send({
            op: "stop",
            guildId: this.guild.id,
        });
    }

    /**
     * Pauses the current track.
     * @param {boolean} pause - Whether to pause the current track.
     */
    public pause(pause: boolean): void {
        if (typeof pause !== "boolean") { throw new RangeError("Player#pause(pause: boolean) Pause can only be \"true\" or \"false\"."); }
        this.playing = !pause;
        this.node.send({
            op: "pause",
            guildId: this.guild.id,
            pause,
        });
    }

    /**
     * Seeks to the position in the current track.
     * @param {boolean} pause - Whether to pause the current track.
     */
    public seek(position: number): void {
        if (!this.queue[0]) { throw new RangeError("Player#seek(position: number) Can only seek when theres a track in the queue."); }
        if (isNaN(position)) { throw new RangeError("Player#seek(position: number) Position must be a number."); }
        if (position < 0 || position > this.queue[0].duration) { throw new RangeError(`Player#seek(position: number) Position can not be smaller than 0 or bigger than ${this.queue[0].duration}.`); }
        this.position = position;
        this.node.send({
            op: "seek",
            guildId: this.guild.id,
            position,
        });
    }
}
