// erela.js/src/lib/structures/Track.ts
import { User } from "discord.js";

/**
 * The ITrackInfo interface.
 */
export interface ITrackInfo {
    /**
     * The track's identifier.
     */
    readonly identifier: string;
    /**
     * Whether the track is seekable.
     */
    readonly isSeekable: boolean;
    /**
     * The author of the track.
     */
    readonly author: string;
    /**
     * The track's length.
     */
    readonly length: number;
    /**
     * Whether the track is a string.
     */
    readonly isStream: boolean;
    /**
     * The track's title.
     */
    readonly title: string;
    /**
     * The track's URI.
     */
    readonly uri: string;
}

/**
 * The ITrackData interface.
 */
export interface ITrackData {
    /**
     * The base 64 encoded track.
     */
    readonly track: string;
    /**
     * The tracks info.
     */
    readonly info: ITrackInfo;
}

/**
 * The Track class.
 * @export
 * @class Track
 */
export class Track {
    /**
     * The base 64 encoded track.
     */
    public track: string;
    /**
     * The track's identifier.
     */
    public identifier: string;
    /**
     * Whether the track is seekable.
     */
    public isSeekable: boolean;
    /**
     * The author of the track.
     */
    public author: string;
    /**
     * The track's duration.
     */
    public duration: number;
    /**
     * Whether the track is a string.
     */
    public isStream: boolean;
    /**
     * The track's title.
     */
    public title: string;
    /**
     * The track's URI.
     */
    public uri: string;
    /**
     * The track's thumbnail.
     */
    public thumbnail: string;
    /**
     * The user who requested the track.
     */
    public requester: User;
    /**
     * Creates an instance of Track.
     * @param {ITrackData} data - The data to pass.
     * @param {User} user - The user who requested the track.
     */
    public constructor(data: ITrackData, user: User) {
        if (!data || !user) {
            throw new RangeError("Track constructor must have all two parameters filled.");
        }

        try {
            this.track = data.track;
            this.identifier = data.info.identifier;
            this.isSeekable = data.info.isSeekable;
            this.author = data.info.author;
            this.duration = data.info.length;
            this.isStream = data.info.isStream;
            this.title = data.info.title;
            this.uri = data.info.uri;
            this.thumbnail = this.uri.includes("youtube") ?
                `https://img.youtube.com/vi/${this.identifier}/default.jpg` : "";
            this.requester = user;
        } catch (err) {
            throw new RangeError(`Invalid track passed. Reason: ${err}`);
        }
    }
}
